revocation/crl/cache/file.go

```diff
@@ -0,0 +1,272 @@
+	return os.Rename(tempFile.Name(), filepath.Join(c.root, fileName(uri)))
+}
+
+// Delete removes the CRL bundle file from file system
+func (c *FileCache) Delete(ctx context.Context, uri string) error {
+	// remove is atomic on UNIX-like platforms
+	return os.Remove(filepath.Join(c.root, fileName(uri)))
+}
+
+// Flush removes all CRLs from the file system
+func (c *FileCache) Flush(ctx context.Context) error {
+	return filepath.Walk(c.root, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+
+		// remove is atomic on UNIX-like platforms
+		return os.Remove(path)
+	})
+}
+
+// fileName returns the file name of the CRL bundle tarball
+func fileName(url string) string {
+	return hashURL(url) + ".tar"
+}
+
+// hashURL hashes the URL with SHA256 and returns the hex-encoded result
+func hashURL(url string) string {
+	hash := sha256.Sum256([]byte(url))
+	return hex.EncodeToString(hash[:])
+}
+
+// parseBundleFromTar parses the CRL blob from a tarball
+//
+// The tarball should contain two files:
+// - base.crl: the base CRL in DER format
+// - metadata.json: the metadata of the CRL
+//
+// example of metadata.json:
+//
+//	{
+//	  "base.crl": {
+//	    "url": "https://example.com/base.crl"
+//	  },
+//	  "createAt": "2024-07-20T00:00:00Z"
+//	}
+func parseBundleFromTar(data io.Reader) (*Bundle, error) {
+	bundle := &Bundle{}
+
+	// parse the tarball
+	tar := tar.NewReader(data)
+	for {
+		header, err := tar.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, &BrokenFileError{
+				Err: fmt.Errorf("failed to read tarball: %w", err),
+			}
+		}
+
+		switch header.Name {
+		case PathBaseCRL:
+			// parse base.crl
+			data, err := io.ReadAll(tar)
+			if err != nil {
+				return nil, err
+			}
+
+			var baseCRL *x509.RevocationList
+			baseCRL, err = x509.ParseRevocationList(data)
+			if err != nil {
+				return nil, &BrokenFileError{
+					Err: fmt.Errorf("failed to parse base CRL from tarball: %w", err),
+				}
+			}
+			bundle.BaseCRL = baseCRL
+		case PathMetadata:
+			// parse metadata
+			var metadata Metadata
+			if err := json.NewDecoder(tar).Decode(&metadata); err != nil {
+				return nil, &BrokenFileError{
+					Err: fmt.Errorf("failed to parse CRL metadata from tarball: %w", err),
+				}
+			}
+			bundle.Metadata = metadata
+		}
+	}
+	if err := bundle.Validate(); err != nil {
+		return nil, err
+	}
+
+	return bundle, nil
+}
+
+// SaveAsTar saves the CRL blob as a tarball, including the base CRL and
+// metadata
+//
+// The tarball should contain two files:
+// - base.crl: the base CRL in DER format
+// - metadata.json: the metadata of the CRL
+//
+// example of metadata.json:
+//
+//	{
+//	  "base.crl": {
+//	    "url": "https://example.com/base.crl"
+//	  },
+//	  "createAt": "2024-06-30T00:00:00Z"
+//	}
+func saveTar(w io.Writer, bundle *Bundle) (err error) {
+	if err := bundle.Validate(); err != nil {
+		return err
+	}
```

Comment:
Should this be checked before creating the temp file?

Reply:
Updated.